File: apps/android/app/src/main/java/dev/flextranslate/foundation/GeminiFlashConfig.kt
package dev.flextranslate.foundation

/**
 * Настройки облачного MT-тира Gemini Flash (`gemini-flash-cloud`, WS5).
 *
 * Конкретный id модели Gemini держим в конфиге, а не зашиваем в код (docs/design/ws5-gemini-flash.md §4):
 * так "3.5 Flash" или что-то поновее можно поменять без правки кода.
 *
 * Два режима работы (см. [GeminiCredentialMode]):
 *
 * **[GeminiCredentialMode.BACKEND_MEDIATION]** (исходный путь WS5): тут endpoint'ы НАШЕГО бэкенда,
 * а не Google. Приложение никогда не держит host + ключ Google вместе и не везёт с собой ключ
 * Gemini. Реальный ключ подставляет бэкенд на своей стороне и возвращает уже переведённый текст.
 *
 * **[GeminiCredentialMode.OWN_KEY]** (BYOK): пользователь даёт свой ключ Gemini, он хранится
 * зашифрованным на устройстве (EncryptedSharedPreferences). Приложение бьёт прямо в публичный REST
 * Gemini. Работает там, где Gemini не заблокирован по гео; если регион не поддерживается — честно
 * показываем это (HTTP 400 "User location is not supported").
 *
 * @property modelId id модели Gemini. По умолчанию `gemini-3.5-flash` (GA 2026-05-19).
 * @property credentialMode через бэкенд-посредник или со своим ключом.
 * @property backendBaseUrl базовый URL бэкенда-посредника. Пусто → гейт честно блокирует с причиной
 *   "Не указан backend-endpoint" (актуально только для BACKEND_MEDIATION).
 * @property mediatedTranslatePath путь на бэкенде для вызова перевода через посредника.
 * @property thinkingLevel `low` ради скорости — перевод чувствителен к задержке. Прокидываем на бэкенд/Gemini.
 * @property streaming вариант с SSE-стримингом; задел под будущий диалоговый поток.
 * @property timeoutMs жёсткий сетевой таймаут. Никогда не оставлять бесконечный дефолт (правило безопасности).
 */
data class GeminiFlashConfig(
    val modelId: String = DEFAULT_MODEL_ID,
    val credentialMode: GeminiCredentialMode = GeminiCredentialMode.BACKEND_MEDIATION,
    val backendBaseUrl: String = "",
    val mediatedTranslatePath: String = "/v1/cloud/translate",
    val liveTokenPath: String = "/v1/cloud/live-token",
    val thinkingLevel: String = "low",
    val streaming: Boolean = false,
    val timeoutMs: Int = DEFAULT_TIMEOUT_MS,
) {
    /** True, только если задан непустой базовый URL бэкенда (без него BACKEND_MEDIATION не запустить). */
    val hasBackend: Boolean get() = backendBaseUrl.isNotBlank()

    /** Полный URL endpoint'а перевода или null, если бэкенд не настроен. */
    fun translateEndpoint(): String? {
        if (!hasBackend) return null
        val base = backendBaseUrl.trimEnd('/')
        val path = if (mediatedTranslatePath.startsWith('/')) mediatedTranslatePath else "/$mediatedTranslatePath"
        return base + path
    }

    companion object {
        /** Текущая быстрая GA-модель Gemini (GA 2026-05-19). Дефолт из конфига, апгрейдится без правки кода. */
        const val DEFAULT_MODEL_ID = "gemini-3.5-flash"
        const val DEFAULT_TIMEOUT_MS = 15_000
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/GeminiFlashConfig.swift
import Foundation

/// Config for the WS5 Gemini Flash cloud MT tier.
///
/// The concrete Gemini model id is config, never a literal, so it can change with no code edit.
///
/// Two modes are supported (see GeminiCredentialMode):
///
/// backendMediation (original path): translation goes through OUR backend, which injects the
/// Gemini credential server-side and returns translated text only. The app never holds a Gemini key.
///
/// ownKey (BYOK): the user supplies their own Gemini API key, stored in the iOS Keychain.
/// The app POSTs directly to the public Gemini REST endpoint. Geo-restriction surfaced honestly.
///
/// Mirrors Android GeminiFlashConfig.
struct GeminiFlashConfig: Sendable {
    let modelId: String
    let credentialMode: GeminiCredentialMode
    let backendBaseUrl: String
    let mediatedTranslatePath: String
    let thinkingLevel: String
    let streaming: Bool
    let timeoutSeconds: Double

    init(
        modelId: String = Self.defaultModelId,
        credentialMode: GeminiCredentialMode = .backendMediation,
        backendBaseUrl: String = "",
        mediatedTranslatePath: String = "/v1/cloud/translate",
        thinkingLevel: String = "low",
        streaming: Bool = false,
        timeoutSeconds: Double = 15
    ) {
        self.modelId = modelId
        self.credentialMode = credentialMode
        self.backendBaseUrl = backendBaseUrl
        self.mediatedTranslatePath = mediatedTranslatePath
        self.thinkingLevel = thinkingLevel
        self.streaming = streaming
        self.timeoutSeconds = timeoutSeconds
    }

    /// Current GA fast Gemini model. Upgradable without code change.
    static let defaultModelId = "gemini-3.5-flash"

    /// True only when a non-blank backend base URL is configured.
    var hasBackend: Bool { !backendBaseUrl.trimmingCharacters(in: .whitespaces).isEmpty }

    /// Absolute translate endpoint, or nil when no backend is configured.
    func translateEndpoint() -> String? {
        guard hasBackend else { return nil }
        let base = backendBaseUrl.hasSuffix("/")
            ? String(backendBaseUrl.dropLast())
            : backendBaseUrl
        let path = mediatedTranslatePath.hasPrefix("/")
            ? mediatedTranslatePath
            : "/\(mediatedTranslatePath)"
        return base + path
    }
}
